### ISSUE TITLE
split lines with point layer

### ISSUE THREAD

**Issue by awa5114** [geopandas/geopandas#1356]
This is basic GIS functionality which, in my opinion should be available out of the box in `geopandas`. `shapely` [offers this functionality at a basic geometric level an](https://shapely.readthedocs.io/en/latest/manual.html#splitting)d an eventual implementation could resemble set overlay methods:

```
import geopandas

points = geopandas.read_file('points.shp')
lines = geopandas.read_file('lines.shp')

segments = geopandas.split(lines, points, tol=5)
```

What remains to be determined is how the attributes should be handled during such a split. If my memory is correct, common GIS software will usually keep existing geometric/attribute discontinuities and further apply the split-by-point operation on top of this...

Any ideas? Would such a feature be desirable?

**Comment by martinfleis** [comment]
Absolutely. I have wrote a version of this for GeoSeries earlier, posting for a reference. If you want to implement it, you are more than welcome to do so. Shapely will return `GeometryCollection`, so there is no issue with preserving attributes. 

```py
def _split(lines, points, tolerance):
    lines = lines.copy()
    buf = points.buffer(tolerance)
    b_sindex = buf.sindex
    for i, geom in lines.iteritems():
        bounds = list(geom.bounds)
        bounds[0] -= tolerance
        bounds[1] -= tolerance
        bounds[2] += tolerance
        bounds[3] += tolerance
        poss_ix = list(b_sindex.intersection(bounds))
        poss_m = buf.iloc[poss_ix]
        s = poss_m.intersects(geom)
        intersects = points.loc[s[s].index]
        if not intersects.empty:
            geom = shapely.ops.split(geom, intersects.unary_union)

            input.loc[i] = geom

    return input
```

**Comment by brendan-ward** [comment]
Alternative implementation [here](https://github.com/brendan-ward/nhdnet/blob/master/nhdnet/geometry/lines.py#L204-L261) which assumes, but does not require, that points are snapped to lines in a separate [step](https://github.com/brendan-ward/nhdnet/blob/master/nhdnet/geometry/lines.py#L49-L128).  (in my case, I needed to preserve topology of split lines and there could be multiple points per line).

A few points to consider, which I approached differently than example above:
* you can create the spatial index of the lines instead of the points
* you can create a "window" around each point by +/- tolerance to [query](https://github.com/brendan-ward/nhdnet/blob/master/nhdnet/geometry/lines.py#L86-L88) the `sindex` on lines, then toss out any lines that are beyond the tolerance distance.  This can be vectorized and reasonably fast; I found that switching from a loop to a vectorized approach saved me considerable processing time.  Depending on your density of lines vs points, this would likely be faster than first buffering points.
* You can use shapely's `project` and `interpolate` functions to snap the point to the nearest line before cutting the line at each of those snapped points using whatever method is best; I built them up using `interpolate` and manual creation of `LineString`s, but a union operation could potentially work too.  It's been a while so I don't recall all the details, but I remember edge cases with union if the point occurred precisely at an endpoint of a line.  For splitting lines, these points should be tossed out first since they don't actually split the lines; in my approach I needed them because I was locating points on a linear network.

In case it's useful for understanding the snapping approach described above, I wrote [this](https://medium.com/@brendan_ward/how-to-leverage-geopandas-for-faster-snapping-of-points-to-lines-6113c94e59aa) a while back.  I've since shifted to using `pygeos` for snapping (using an approximation of a nearest neighbor approach on top of `pygeos::STRtree`) but haven't had a chance to write that up yet.


@martinfleis am I reading your example correctly in that the points would first need to be precisely snapped to the lines in order for `unary_union` to function properly?

**Comment by martinfleis** [comment]
> am I reading your example correctly in that the points would first need to be precisely snapped to the lines in order for unary_union to function properly?

@brendan-ward Yes. This was never finished, so it is a bit rough. I think that the initial idea was to have `snap` along the way to actually make sense of `tolerance`. But using unary_union you avoid the need of looping through the geometry if it should be split on multiple points.

**Comment by scoddou** [comment]
@brendan-ward I stumbled on this chain when looking for a way to slice all crossing linestrings in a single gdf. Not exactly related to your question, but wanted to mention in case you found it useful. Essentially, I wanted to find a way to leverage spatial indexing available in ```sjoin```.  

Just finally jumping back into ```geopandas``` since the upgrade to ```shapely 2.0```, so code below might be a bit messy

```python
import geopandas as gpd
from shapely.geometry import LineString

line1 = LineString([(0, 0), (1000, 1000)])
line2 = LineString([(0, 1000), (1000, 0)])
line3 = LineString([(0, 1000), (1000, 500)])
lines = gpd.GeoDataFrame({'id': [20,21,22], 'other':[50,51,52]}, geometry=[line1,line2,line3])

slicer = gpd.sjoin(lines, lines, how='right', predicate='crosses')
lines.set_index('id', inplace=True, drop=False)
slicer.set_index('id_left', inplace=True, drop=True)
slicer = slicer.dissolve(by='id_left').geometry.reindex(lines.index, fill_value=LineString())

sub_lines = lines.difference(slicer, align=True).explode(index_parts=False)

gdf = gpd.GeoDataFrame(lines.drop(columns='geometry').loc[sub_lines.index], geometry=sub_lines)

# optional index reset
# gdf = gdf.reset_index(drop=True)

**Comment by MTueting** [comment]
> @brendan-ward I stumbled on this chain when looking for a way to slice all crossing linestrings in a single gdf. Not exactly related to @amine-aboufirass question, but wanted to mention in case anyone found it useful. Essentially, I wanted to find a way to leverage spatial indexing available in `sjoin`.

Incredible solution! Thank you so much. Was tinkering around with much more complicated code (grab all coordinates of the original lines, unary_union to blend the lines, which also makes the new nodes at intersections, then grab the new coordinates and split the lines at these.), but this is much faster and convenient! Cheers
